jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DoSHandler.java

```diff
@@ -181,44 +181,42 @@ public void setServer(Server server)
     @Override
     protected boolean onConditionsMet(Request request, Response response, Callback callback) throws Exception
     {
-        // Reject if we have too many Trackers
-        if (_maxTrackers > 0 && _trackers.size() >= _maxTrackers)
-        {
-            // Try shrinking the tracker pool
-            long now = NanoTime.now();
-            _trackers.values().removeIf(tracker -> tracker.isIdle(now));
-            if (_trackers.size() >= _maxTrackers)
-            {
-                // Try shrinking the tracker pool as if we are at the next sample period already
-                long nextIdleCheck = NanoTime.now() + _trackerFactory.getSamplePeriod().getNano();
-                _trackers.values().removeIf(tracker -> tracker.isIdle(nextIdleCheck));
-                if (_trackers.size() >= _maxTrackers)
-                    return _rejectHandler.handle(request, response, callback);
-            }
-        }
+        // use NanoTime.now() instead of request.getBeginNanoTime() to avoid jitter
+        long now = NanoTime.now();
 
-        // Calculate an id for the request (which may be global empty string)
-        String id = _getId.apply(request);
+        // Calculate an id for the request (which may be global empty string).
+        String id = _clientIdFn.apply(request);
 
+        // Reject or handle untracked request
         if (id == null)
-            return _rejectHandler.handle(request, response, callback);
+            id = "";
 
-        // Obtain a tracker, creating a new one if necessary.  Trackers are removed if CyclicTimeouts#onExpired returns true
+        // Obtain a tracker, creating a new one if necessary (and not too many)
+        // Trackers are removed if CyclicTimeouts#onExpired returns true.
         Tracker tracker = _trackers.computeIfAbsent(id, this::newTracker);
 
-        // If we are not over-limit then handle normally
-        if (tracker.onRequest(request.getBeginNanoTime()))
+        // If we have too many trackers, then we will have a null tracker
+        if (tracker == null)
+            return _rejectUntracked ? _rejectHandler.handle(request, response, callback) : nextHandler(request, response, callback);
+
+        // IS the request allowed by the tracker?
+        boolean allowed = tracker.onRequest(now);
```

Comment:
I think `onRequest(long)` should be changed to `onRequest(Request rq)`, so that implementations can examine the request and possibly allow some unconditionally.

Reply:
That will make trackers impossible to test, as they will be taking nanotime internally rather than having it passed in from a unit test.    Everytime we do something like that, we create flaky tests.

Besides, we are already in `onConditionsMet` method, so the request has already been filtered and unconditional requests can be setup in the conditional handler.   There is no need for a) two mechanisms for the same thing; b) being overly generic based on zero use-cases.